feat: migrate to Bun + Hono + TypeScript with dual runtime support

Shahfarzane opened the pull request:
## Summary

This PR migrates the Color Name API to use Bun + Hono with full TypeScript support while maintaining backward compatibility with Node.js.

### Core Changes
- Migrate server to **Bun + Hono** framework with full TypeScript support
- Add **dual runtime support**: Node.js (via tsx) and Bun  
- Maintain backward compatibility with existing JavaScript server (`npm run start`)

### New Features
- Socket.IO live updates with client location broadcasts
- Lazy-load GeoIP for better startup performance
- Improved CORS handling with X-Referrer header support
- Rate limiting middleware with configurable limits
- Comprehensive environment variable configuration

### Documentation
- `.env.example` - Complete environment template
- `docs/CONFIGURATION.md` - Environment variables reference
- `docs/DEPLOYMENT.md` - Fly.io, Docker, VPS deployment guides
- `docs/DEVELOPMENT.md` - Local development setup

### Dual Runtime Scripts

| Task | Node.js (tsx) | Bun |
|------|---------------|-----|
| Dev server | `npm run dev` | `bun run dev:bun` |
| Run tests | `npm run test` | `bun run test:bun` |
| Start TS | `npm run start:ts` | `bun run start:bun` |

### Testing
- Concurrent user tests added
- All tests work with both Node.js and Bun runtimes
- TypeScript test files with tsx support

## Test Plan
- [ ] Run `npm install && npm run test` (Node.js)
- [ ] Run `bun install && bun run test:bun` (Bun)
- [ ] Verify API responses match existing behavior
- [ ] Test Socket.IO live updates

meodai commented:
@Shahfarzane thanks so much for this! Its a lot to process at once. Would it be possible to not touch gh-pages for now? Because I will continue to run it on heroku for a little while until we figure out a good strategy to move all that.

Since its all TS now I can not just cherry pick.

Shahfarzane committed:
feat: migrate to Bun + Hono + TypeScript with dual runtime support

## Core Changes
- Migrate server to Bun + Hono framework with full TypeScript support
- Add dual runtime support: Node.js (via tsx) and Bun
- Maintain backward compatibility with existing JavaScript server

## New Features
- Socket.IO live updates with client location broadcasts
- Lazy-load GeoIP for better startup performance
- Improved CORS handling with X-Referrer header support
- Rate limiting middleware with configurable limits

## Documentation
- Add comprehensive ENV variable documentation (.env.example)
- Add docs/CONFIGURATION.md - complete environment reference
- Add docs/DEPLOYMENT.md - Fly.io, Docker, VPS deployment guides
- Add docs/DEVELOPMENT.md - local development setup

## Testing
- Add concurrent user tests
- Add TypeScript test files with tsx support
- Tests work with both Node.js and Bun runtimes

## Scripts
- Default scripts use tsx (Node-compatible)
- :bun suffix scripts for Bun users
- npm run dev / bun run dev:bun
- npm run test / bun run test:bun

Note: gh-pages changes excluded per maintainer request

Shahfarzane commented:
Updated! I've removed all `gh-pages/` changes from this PR. The PR now only includes:

- Core server migration (Bun + Hono + TypeScript)
- Dual runtime support (Node.js via tsx + Bun)
- Documentation (docs/, .env.example)
- Tests
- Configuration files (fly.toml, Dockerfile.bun, etc.)

The gh-pages folder is now untouched and stays exactly as it was. Let me know if you need any other changes!

meodai commented:
Thanks for removing the gh-pages changes! That helps a lot. 🙏

I'm still a bit overwhelmed by the scope though: this is essentially a complete rewrite touching 80+ files. It's hard for me to get a grasp of the concrete changes.

Would you be open to breaking this into a few smaller PRs? Something like:

TypeScript migration only: convert existing JS files to TS with minimal logic changes
Bun + Hono migration: swap the framework/runtime
New features: Socket.IO live updates, rate limiting, etc.

I totally understand if that feels like a lot of rework, but it would really help me review with confidence. The TypeScript conversion especially makes it tricky because I can't easily diff what actually changed behavior-wise vs. what's just type annotations.

What do you think? Happy to discuss if there's a middle ground that works for both of us.
I love a lot of the ideas I saw, like presenting the user with an index on the base URL.

meodai commented on .gitignore:
```diff
@@ -1,2 +1,36 @@
+# typescript
+*.tsbuildinfo
+next-env.d.ts
+gh-pages/.DS_Store
```
this is redudant with the line above .DS_Store and should be part of the gobal git ignore imho
